### PR TITLE
Golang: Debugging memory leaks using pprof

### PR DESCRIPTION
**Pull request by davidsbond**

### CODE REVIEW

File: _posts/2018-08-08-debugging-memory-leaks-using-pprof.md [hunk 1]
```diff
@@ -0,0 +1,218 @@
+---
+layout: post
+title:  "Golang: Debugging memory leaks using pprof"
+date:   2018-08-08
+tags: golang http debugging pprof memory leak
+---
+
+## Introduction 
+
+I work as a software engineer at [OVO Energy](https://www.ovoenergy.com/) where my team are implementing the CRM solution used by customer services. We've currently been building a new set of microservices to eventually replace the existing one. One of our microservices in responsible for migrating data from the old system into the new one.
```
Comment (mickwheelz) [inline _posts/2018-08-08-debugging-memory-leaks-using-pprof.md]: "We're currently building"

"replace the existing services"

"one of our microservices is responsible"

File: _posts/2018-08-08-debugging-memory-leaks-using-pprof.md [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+A few days after deploying a new version of the service, I opened the relevant monitoring dashboard and saw this:
+
+![Memory usage graph]({{ "/assets/2018-08-08-debugging-memory-leaks-using-pprof/1.png" | absolute_url }})
+
+According to this graph, we have a memory leak somewhere. This is most likely due to an issue with the management of goroutines within the service. However, the service relies heavily on concurrency and finding where the leak is might not be so easy. Luckily, goroutines are lightweight, allowing a reasonable amount of time to figure out where the leak is before it becomes a real/expensive problem. The two spikes on the 12pm marks are times when migrations occurred.
+
+### Background
+
+Over the course of a few weeks I designed and implemented the service and hosted it in our [Kubernetes](https://kubernetes.io) cluster on [GCP](https://cloud.google.com/), ensuring to add monitoring functionality in order to make it ready for production. This included an HTTP endpoint for health checks, log-based metrics and uptime checks using [Stackdriver](https://cloud.google.com/stackdriver/).
```
Comment (mickwheelz) [inline _posts/2018-08-08-debugging-memory-leaks-using-pprof.md]: "ensuring that I added monitoring"

File: _posts/2018-08-08-debugging-memory-leaks-using-pprof.md [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    "kafka": {
+        "status": "UP"
+    },
+    "registry": {
+        "status": "UP"
+    }
+}
+```
+
+First thing I would note, using `runtime.NumGoroutine()` to see the number of running goroutines is extremely helpful identifying the source of the memory leak. I recommend having some way to monitor this in your production environments. In this scenario, our HTTP health check returns this value.
```
Comment (mickwheelz) [inline _posts/2018-08-08-debugging-memory-leaks-using-pprof.md]: "extremely helpful in identifying"

"In this scenario, our HTTP health check returns the number of running goroutines" - maybe for clarity?

File: _posts/2018-08-08-debugging-memory-leaks-using-pprof.md [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+On the day of the leak, I saw the number of goroutines exceed 100000 and keep rising steadily with each health check request. Below are the steps I took in debugging this issue.
+
+### Enabling pprof output via HTTP
+
+The `pprof` tool describes itself as "a tool for visualization and analysis of profiling data", you can view the GitHub repository for it [here](https://github.com/google/pprof). This tool allows us to obtain various metrics on the low-level operations of a Go program. For our purposes, it allows us to get detailed information on running goroutines. The only problem here is that `pprof` is a binary we would have to run commands against the service in production for meaningful results. The application also runs within a [Docker](https://www.docker.com/) container based on a `scratch` image, which makes using the binary somewhat invasive. How then can we get the profiling data we need?
```
Comment (mickwheelz) [inline _posts/2018-08-08-debugging-memory-leaks-using-pprof.md]: "is that `pprof` is a binary, we would"

File: _posts/2018-08-08-debugging-memory-leaks-using-pprof.md (hunk 1 above)
Comment (chriswaldron-ovo) [inline _posts/2018-08-08-debugging-memory-leaks-using-pprof.md]: One of our microservices in responsible for migrating data from the old system into the new one.
-> One of our microservices is responsible for migrating data from the old system into the new one.

File: _posts/2018-08-08-debugging-memory-leaks-using-pprof.md [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+## Introduction 
+
+I work as a software engineer at [OVO Energy](https://www.ovoenergy.com/) where my team are implementing the CRM solution used by customer services. We've currently been building a new set of microservices to eventually replace the existing one. One of our microservices in responsible for migrating data from the old system into the new one.
+
+A few days after deploying a new version of the service, I opened the relevant monitoring dashboard and saw this:
+
+![Memory usage graph]({{ "/assets/2018-08-08-debugging-memory-leaks-using-pprof/1.png" | absolute_url }})
+
+According to this graph, we have a memory leak somewhere. This is most likely due to an issue with the management of goroutines within the service. However, the service relies heavily on concurrency and finding where the leak is might not be so easy. Luckily, goroutines are lightweight, allowing a reasonable amount of time to figure out where the leak is before it becomes a real/expensive problem. The two spikes on the 12pm marks are times when migrations occurred.
```
Comment (chriswaldron-ovo) [inline _posts/2018-08-08-debugging-memory-leaks-using-pprof.md]: However, the service relies heavily on concurrency and finding where the leak is might not be so easy
-> However, the service relies heavily on concurrency, and finding where the leak is might not be so easy

(I had to re-read it without the comma but I am a slow boi)

File: _posts/2018-08-08-debugging-memory-leaks-using-pprof.md (hunk 3 above)
Comment (chriswaldron-ovo) [inline _posts/2018-08-08-debugging-memory-leaks-using-pprof.md]: First thing I would note, using runtime.NumGoroutine() to see the number of running goroutines is extremely helpful identifying the source of the memory leak

-> First thing I would note, using runtime.NumGoroutine() to see the number of running goroutines is extremely helpful *in* identifying the source of the memory leak

**Comment by davidsbond** [comment]
@mickwheelz @chriswaldron-ovo Have made changes based on your feedback <3

**Pull request merged**